HiP-ThumbnailService/Controllers/ThumbnailsController.cs

```diff
@@ -0,0 +1,160 @@
+
+            var semaphore = LockDictionary.GetOrAdd(encodedId, new SemaphoreSlim(1));
+
+            await semaphore.WaitAsync();
+            try
+            {
+                //clear folder
+                if (Directory.Exists(folderPath))
+                {
+                    var folder = new DirectoryInfo(folderPath);
+                    folder.Delete(true);
+                }
+                
+                return NoContent();
+            }
+            finally
+            {
+                semaphore.Release();
+            }
+        }
+
+        [ProducesResponseType(400)]
+        [ProducesResponseType(200)]
+        [HttpGet]
+        public async Task<IActionResult> Get(CreationArgs args)
+        {
+            if (!ModelState.IsValid)
+                return BadRequest(ModelState);
+
+            if (string.IsNullOrEmpty(args.Url)) return BadRequest(new { Message = "An url for image access must be provided" });
+
+            var encodedId = Convert.ToBase64String(Encoding.UTF8.GetBytes(args.Url));
+            var folderPath = Path.Combine(_dirConfig.Path, encodedId);
+
+            if (string.IsNullOrEmpty(args.Size) || !_sizeConfig.SupportedSizes.ContainsKey(args.Size))
+                return BadRequest(new { Message = "Invalid size" });
+
+            var semaphore = LockDictionary.GetOrAdd(encodedId, new SemaphoreSlim(1));
+            await semaphore.WaitAsync();
+            try
+            {
+                var extension = RequestImageFormatUtils.GetExtension(args.Format);
+                var thumbnailPath = Path.Combine(folderPath, GetFileName(args.Size, args.Mode)) + "." +
+                                    extension;
+                if (!System.IO.File.Exists(thumbnailPath))
+                {
+                    var client = new HttpClient();
+                    client.DefaultRequestHeaders.Add("Authorization", Request.Headers["Authorization"].ToString());
+                    var hostUrl = _endpointConfig.HostUrl;
+                    var stream = await client.GetStreamAsync(hostUrl + args.Url);
+
+                    if (stream == null)
+                    {
+                        return BadRequest(new { Message = "Image access failed" });
+                    }
+
+                    //Create Directory if it doesn't exist
+                    if (!Directory.Exists(folderPath))
+                        Directory.CreateDirectory(folderPath);
+
+                    var value = _sizeConfig.SupportedSizes[args.Size];
+                    GenerateThumbnail(args.Mode, stream, thumbnailPath, value);
+                }
+
+                return File(new FileStream(thumbnailPath, FileMode.Open), $"image/{extension}",
+                        Path.GetFileName(thumbnailPath));
+
+            }
+            finally
+            {
+                semaphore.Release();
+            }
+
+        }
+
+        /// <summary>
+        /// Generates a thumbnail, if it doesn't exist yet
```

Comment:
Wrong (probably outdated?) doc: The method *always* generates a thumbnail, even if it already exists.